In [23]:
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import KFold
from sklearn.metrics import r2_score

1. Загрузите  данные из файла abalone.csv. Это датасет,в котором тре-буется предсказать возраст ракушки (число колец) по физическим
измерениям.

In [8]:
data = pd.read_csv('abalone.csv')
print(data)

     Sex  Length  Diameter  Height  WholeWeight  ShuckedWeight  VisceraWeight  \
0      M   0.455     0.365   0.095       0.5140         0.2245         0.1010   
1      M   0.350     0.265   0.090       0.2255         0.0995         0.0485   
2      F   0.530     0.420   0.135       0.6770         0.2565         0.1415   
3      M   0.440     0.365   0.125       0.5160         0.2155         0.1140   
4      I   0.330     0.255   0.080       0.2050         0.0895         0.0395   
...   ..     ...       ...     ...          ...            ...            ...   
4172   F   0.565     0.450   0.165       0.8870         0.3700         0.2390   
4173   M   0.590     0.440   0.135       0.9660         0.4390         0.2145   
4174   M   0.600     0.475   0.205       1.1760         0.5255         0.2875   
4175   F   0.625     0.485   0.150       1.0945         0.5310         0.2610   
4176   M   0.710     0.555   0.195       1.9485         0.9455         0.3765   

      ShellWeight  Rings  


2. Преобразуйте признак Sex в числовой: значение F должно перейти
в -1, I — в 0, M — в 1. Если вы используете Pandas, то подойдет следующий код: data[’Sex’] = data[’Sex’].map(lambda x: 1 if x ==
’M’ else (-1 if x == ’F’ else 0))

In [9]:
data['Sex'] = data['Sex'].map(
    lambda x: 1 if x == 'M' else (-1 if x == 'F' else 0)
)
print(data['Sex'])

0       1
1       1
2      -1
3       1
4       0
       ..
4172   -1
4173    1
4174    1
4175   -1
4176    1
Name: Sex, Length: 4177, dtype: int64


3. Разделите содержимое файлов на признаки и целевую переменную.
В последнем столбце записана целевая переменная, в остальных —
признаки.

In [20]:
X = data.drop(columns='Rings')
y = data['Rings']

4. Обучите случайный лес (sklearn.ensemble.RandomForestRegressor)
с различным числом деревьев: от 1 до 50 (random_state=1). Для
каждого из вариантов оцените качество работы полученного леса
на кросс-валидации по 5 блокам. Используйте параметры
"random_state=1"и "shuffle=True"при создании генератора кросс-
валидацииsklearn.cross_validation.KFold.В качестве меры качества
воспользуйтесь коэффициентом детерминации(sklearn.metrics.r2_score).

In [27]:
cv = KFold(n_splits=5, shuffle=True, random_state=1)
best_score = -10
best_i = 0
for i in range(1, 51):
    scores = []
    for train_i, test_i in cv.split(X):
        X_train, X_test = X.iloc[train_i], X.iloc[test_i]
        y_train, y_test = y.iloc[train_i], y.iloc[test_i]
        
        clf = RandomForestRegressor(n_estimators=i, random_state = 1)
        clf.fit(X_train, y_train)
        pred = clf.predict(X_test)
        score = r2_score(y_test, pred)
        scores.append(score)

    mean_score = np.mean(scores)
    
    if mean_score > best_score:
        best_score = mean_score
        best_i = i
print(best_i)
print(best_score)

50
0.5309509147417047


5. Определите, при каком минимальном количестве деревьев случай-
ный лес показывает качество на кросс-валидации выше 0.52. Это
количество и будет ответом на задание.

In [28]:
cv = KFold(n_splits=5, shuffle=True, random_state=1)
best_score = -10
best_i = 0
for i in range(1, 51):
    scores = []
    for train_i, test_i in cv.split(X):
        X_train, X_test = X.iloc[train_i], X.iloc[test_i]
        y_train, y_test = y.iloc[train_i], y.iloc[test_i]
        
        clf = RandomForestRegressor(n_estimators=i, random_state = 1)
        clf.fit(X_train, y_train)
        pred = clf.predict(X_test)
        score = r2_score(y_test, pred)
        scores.append(score)

    mean_score = np.mean(scores)
    
    if mean_score > 0.52:
        best_score = mean_score
        best_i = i
        break
print(best_i)
print(best_score)

21
0.520529096463528


6. Обратите внимание на изменение качества по мере роста числа де-
ревьев. Ухудшается ли оно?

In [29]:
cv = KFold(n_splits=5, shuffle=True, random_state=1)
best_score = -10
best_i = 0
for i in range(1, 51):
    scores = []
    for train_i, test_i in cv.split(X):
        X_train, X_test = X.iloc[train_i], X.iloc[test_i]
        y_train, y_test = y.iloc[train_i], y.iloc[test_i]
        
        clf = RandomForestRegressor(n_estimators=i, random_state = 1)
        clf.fit(X_train, y_train)
        pred = clf.predict(X_test)
        score = r2_score(y_test, pred)
        scores.append(score)

    mean_score = np.mean(scores)
    print(f'Кол-во деревьев {i} -> качество = {mean_score}')
    

Кол-во деревьев 1 -> качество = 0.10967482068860261
Кол-во деревьев 2 -> качество = 0.3413000096365689
Кол-во деревьев 3 -> качество = 0.406433829066129
Кол-во деревьев 4 -> качество = 0.4447745857536912
Кол-во деревьев 5 -> качество = 0.46503241426823594
Кол-во деревьев 6 -> качество = 0.47139595825898917
Кол-во деревьев 7 -> качество = 0.4766658451893487
Кол-во деревьев 8 -> качество = 0.4829348394224631
Кол-во деревьев 9 -> качество = 0.4894370162945041
Кол-во деревьев 10 -> качество = 0.4954085552428177
Кол-во деревьев 11 -> качество = 0.4944111155773555
Кол-во деревьев 12 -> качество = 0.49902817866563326
Кол-во деревьев 13 -> качество = 0.5030578549564464
Кол-во деревьев 14 -> качество = 0.5073168234618861
Кол-во деревьев 15 -> качество = 0.5091809969556578
Кол-во деревьев 16 -> качество = 0.5114105314179662
Кол-во деревьев 17 -> качество = 0.5148917747729636
Кол-во деревьев 18 -> качество = 0.5172203573170132
Кол-во деревьев 19 -> качество = 0.5198293095329432
Кол-во деревьев 20

Мы видим, что прииерно до 25 дерева заметен активный рост в качестве, дальше с увеличением числа деревьев качество предсказаний приемущественно растет, но незначительно, однако в некоторых случаях качество "fluctuate" - то немного растет, то падает, но в итоге к 48-50 дереву качесво уверенно выходит на 0.53.